### PR TITLE
Multi IP-Adapter for Flux pipelines

### LINKED ISSUE #10775: Support multiple IP adapter in Flux
When I pass the weights in the form of [0.4, 0.4], it tells me "Expected list of 19 scales, got 2."
pipe.set_ip_adapter_scale([0.4, 0.4])

### PR DESCRIPTION
**Pull request by guiyrt**
# What does this PR do?

Fixes #10775. Adds support for multiple IP-Adapters on Flux pipelines. For testing, I tried using a single IP-Adapter with 0.5 scale, and then two equal IP-Adapters with 0.25 scale each, which should (and does) produce the same result. ~Basic functionality is there, but I still want to clean-up some parts and add multi ip-adapter scale tests. While #10758 is not merged, I have the typing helping functions here as well.~ All set!

<details>
<summary>Single IP-Adapter inference code</summary>

```python
import torch
from diffusers import FluxPipeline
from diffusers.utils import load_image

pipe: FluxPipeline = FluxPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-dev",
    torch_dtype=torch.bfloat16,
)

pipe.load_ip_adapter(
    "XLabs-AI/flux-ip-adapter",
    weight_name="ip_adapter.safetensors",
    image_encoder_pretrained_model_name_or_path="openai/clip-vit-large-patch14"
)

pipe.set_ip_adapter_scale(0.5)
pipe.enable_sequential_cpu_offload()

ip_adapter_image = load_image("https://huggingface.co/guiyrt/sample-images/resolve/main/astronaut.jpg")

image = pipe(
    width=1024,
    height=1024,
    prompt="A vintage picture of an astronaut in a starry sky",
    generator=torch.manual_seed(42),
    ip_adapter_image=ip_adapter_image
).images[0]

image.save('result_single.jpg')
```

</details>

<details>
<summary>Multi IP-Adapter inference code</summary>

```python
import torch
from diffusers import FluxPipeline
from diffusers.utils import load_image

pipe: FluxPipeline = FluxPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-dev",
    torch_dtype=torch.bfloat16,
)

pipe.load_ip_adapter(
    ["XLabs-AI/flux-ip-adapter", "XLabs-AI/flux-ip-adapter"],
    weight_name=["ip_adapter.safetensors", "ip_adapter.safetensors"],
    image_encoder_pretrained_model_name_or_path="openai/clip-vit-large-patch14"
)

pipe.set_ip_adapter_scale([0.25, 0.25])
pipe.enable_sequential_cpu_offload()

ip_adapter_image = load_image("https://huggingface.co/guiyrt/sample-images/resolve/main/astronaut.jpg")

image = pipe(
    width=1024,
    height=1024,
    prompt="A vintage picture of an astronaut in a starry sky",
    generator=torch.manual_seed(42),
    ip_adapter_image=[ip_adapter_image,ip_adapter_image]
).images[0]

image.save('result_multi.jpg')
```

</details>

 <table>
        <tr>
            <th>main (sanity check)</th>
            <th>single IP-Adapter, 0.5 scale</th>
            <th>dual IP-Adapter, 2x 0.25 scale </th>
        </tr>
        <tr>
            <td></td>
            <td></td>
            <td></td>
        </tr>
    </table>

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [X] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [X] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@hlky @yiyixuxu

### CODE REVIEW

File: src/diffusers/pipelines/flux/pipeline_flux.py [hunk 1]
```diff
@@ -868,14 +868,23 @@ def __call__(
...
+        # TODO: Clarify this section
         if (ip_adapter_image is not None or ip_adapter_image_embeds is not None) and (
             negative_ip_adapter_image is None and negative_ip_adapter_image_embeds is None
         ):
-            negative_ip_adapter_image = np.zeros((width, height, 3), dtype=np.uint8)
+            negative_ip_adapter_image = (
+                [np.zeros((width, height, 3), dtype=np.uint8) for _ in range(len(ip_adapter_image))]
+                if isinstance(ip_adapter_image, list)
+                else np.zeros((width, height, 3), dtype=np.uint8)
+            )
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux.py:879]:
```suggestion
            negative_ip_adapter_image = np.zeros((width, height, 3), dtype=np.uint8)
            if isinstance(ip_adapter_image, list):
                negative_ip_adapter_image = [negative_ip_adapter_image] * len(ip_adapter_image)
```

File: src/diffusers/pipelines/flux/pipeline_flux.py [hunk 2]
```diff
@@ -868,14 +868,23 @@ def __call__(
...
         elif (ip_adapter_image is None and ip_adapter_image_embeds is None) and (
             negative_ip_adapter_image is not None or negative_ip_adapter_image_embeds is not None
         ):
-            ip_adapter_image = np.zeros((width, height, 3), dtype=np.uint8)
+            ip_adapter_image = (
+                [np.zeros((width, height, 3), dtype=np.uint8) for _ in range(len(negative_ip_adapter_image))]
+                if isinstance(negative_ip_adapter_image, list)
+                else np.zeros((width, height, 3), dtype=np.uint8)
+            )
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux.py:887]:
```suggestion
            ip_adapter_image = np.zeros((width, height, 3), dtype=np.uint8)
            if isinstance(negative_ip_adapter_image, list):
                ip_adapter_image = [ip_adapter_image] * len(negative_ip_adapter_image)
```

File: src/diffusers/pipelines/flux/pipeline_flux.py [hunk 3]
```diff
@@ -868,14 +868,23 @@ def __call__(
         else:
             guidance = None
 
+        # TODO: Clarify this section
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux.py:871]:
```suggestion
        # TODO: Clarify this section
```
Will let @yiyixuxu comment, lgtm though

File: src/diffusers/pipelines/flux/pipeline_flux.py (hunk 1 above)
Comment (guiyrt) [reply src/diffusers/pipelines/flux/pipeline_flux.py:879]: My idea as well, I forgot to change before commit, thanks! 😅 but we also need to account if we use `ip_adapter_image_embeds`, I'll add a check for that

File: src/diffusers/pipelines/flux/pipeline_flux.py [hunk 4]
```diff
@@ -868,19 +873,19 @@ def __call__(
...
-        # TODO: Clarify this section
         if (ip_adapter_image is not None or ip_adapter_image_embeds is not None) and (
             negative_ip_adapter_image is None and negative_ip_adapter_image_embeds is None
         ):
-            negative_ip_adapter_image = np.zeros((width, height, 3), dtype=np.uint8)
-            if isinstance(ip_adapter_image, list):
-                negative_ip_adapter_image = [negative_ip_adapter_image] * len(ip_adapter_image)
+            zeros_image = np.zeros((width, height, 3), dtype=np.uint8)
+            negative_ip_adapter_image_embeds = self.encode_image(zeros_image, device, 1)[None, :]
+            negative_ip_adapter_image_embeds = [negative_ip_adapter_image_embeds] * self.transformer.encoder_hid_proj.num_ip_adapters
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux.py:881]:
```suggestion
            negative_ip_adapter_image = np.zeros((width, height, 3), dtype=np.uint8)
            if isinstance(ip_adapter_image, list):
                negative_ip_adapter_image = [negative_ip_adapter_image] * self.transformer.encoder_hid_proj.num_ip_adapters
```

Let's keep it consistent with other pipelines. `num_ip_adapters` works as a multiplier. Technically inferring once then repeating is faster but negligible.

File: src/diffusers/pipelines/flux/pipeline_flux.py [hunk 5]
```diff
@@ -868,19 +873,19 @@ def __call__(
...
+        
         elif (ip_adapter_image is None and ip_adapter_image_embeds is None) and (
             negative_ip_adapter_image is not None or negative_ip_adapter_image_embeds is not None
         ):
-            ip_adapter_image = np.zeros((width, height, 3), dtype=np.uint8)
-            if isinstance(negative_ip_adapter_image, list):
-                ip_adapter_image = [ip_adapter_image] * len(negative_ip_adapter_image)
+            zeros_image = np.zeros((width, height, 3), dtype=np.uint8)
+            ip_adapter_image_embeds = self.encode_image(zeros_image, device, 1)[None, :]
+            ip_adapter_image_embeds = [ip_adapter_image_embeds] * self.transformer.encoder_hid_proj.num_ip_adapters
```
Comment (hlky) [inline src/diffusers/pipelines/flux/pipeline_flux.py:888]:
```suggestion
            ip_adapter_image = np.zeros((width, height, 3), dtype=np.uint8)
            if isinstance(negative_ip_adapter_image, list):
                ip_adapter_image = [ip_adapter_image] * self.transformer.encoder_hid_proj.num_ip_adapters
```

As above.

**Comment by hlky** [comment]
`FluxIPAdapterTesterMixin` can be updated for multi IPAdapters with `IPAdapterTesterMixin` as reference.

https://github.com/huggingface/diffusers/blob/170833c22a79ac09a3e919345124091f5e917cf3/tests/pipelines/test_pipelines_common.py#L286-L377

**Comment by guiyrt** [comment]
> `FluxIPAdapterTesterMixin` can be updated for multi IPAdapters with `IPAdapterTesterMixin` as reference.
> 

Perfect, will do!

Another thing, in https://github.com/huggingface/diffusers/issues/10775#issuecomment-2671222462, when you mentioned multi image support for `flux-ip-adapter-v2` (to be done on a separate PR), you said `ip_adapter_image` would need to support `List[PipelineImageInput]`. You meant the IP-Adapter interface, not the pipeline interface, right? Because atm `ip_adapter_image` on `FluxPipeline` is still `Optional[PipelineImageInput]`, and I think we need to update to `Optional[Union[PipelineImageInput, List[PipelineImageInput]]` (one image per IP-Adapter).

**Comment by hlky** [comment]
`PipelineImageInput` is 
```python
PipelineImageInput = Union[
    PIL.Image.Image,
    np.ndarray,
    torch.Tensor,
    List[PIL.Image.Image],
    List[np.ndarray],
    List[torch.Tensor],
]
```
The `List` is for `one image per IPAdapter` or `one image per ControlNet`.

`List[PipelineImageInput]` would be like in `ControlNetUnion` where each ControlNet can take multiple inputs (with the experimental scale type, not MultiControlNet).

**Comment by guiyrt** [comment]
Ready for review :) I added the ability to mix and match per-layer scale and single scale. For example, you can pass scale for two IP-Adapters as `[list_with_19_values, 0.3]`. You can also pass a single value and it's used for every layer of every IP-Adapter. I also made use of `_is_valid_type` to check types as part of processing scale input, so it doesn't fail somewhere later, with minimal additional complexity.

**Comment by guiyrt** [comment]
@hlky failed `check_code_quality` comes from `docs/source/en/_toctree.yml`

**Comment by hlky** [comment]
@guiyrt It looks like this has introduced a circular dependency. Let's move `_get_detailed_type` and `_is_valid_type` under `diffusers.utils` and create `typing_utils.py` cc @yiyixuxu 
```python
from ..pipelines.pipeline_loading_utils import _get_detailed_type, _is_valid_type
```

**Comment by guiyrt** [comment]
Just sharing some cool outputs I got, creating image variations by pushing noise through the IP-Adapter using random image embeds instead of outputs from CLIP vision model. Somewhat by accident while testing for `ip_adapter_image_embeds`, but I really liked the vibe from lucky seed 423 🤩 

<details>
<summary>Inference code</summary>

 <details>
<summary>No IP-Adapter</summary>

```python
import torch
from diffusers import FluxPipeline
from diffusers.utils import load_image

pipe: FluxPipeline = FluxPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-dev",
    torch_dtype=torch.bfloat16,
)

pipe.enable_sequential_cpu_offload()

image = pipe(
    width=1024,
    height=1024,
    prompt="A vintage picture of an astronaut in a starry sky",
    generator=torch.manual_seed(423),
).images[0]

image.save('result_no_ipa.jpg')
```
</details>

<details>
<summary>Single IP-Adapter</summary>

```python
import torch
from diffusers import FluxPipeline
from diffusers.utils import load_image

pipe: FluxPipeline = FluxPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-dev",
    torch_dtype=torch.bfloat16,
)

pipe.load_ip_adapter(
    "XLabs-AI/flux-ip-adapter",
    weight_name="ip_adapter.safetensors",
    image_encoder_pretrained_model_name_or_path="openai/clip-vit-large-patch14"
)

pipe.set_ip_adapter_scale(0.5)
pipe.enable_sequential_cpu_offload()

image = pipe(
    width=1024,
    height=1024,
    prompt="A vintage picture of an astronaut in a starry sky",
    generator=torch.manual_seed(423),
    ip_adapter_image_embeds=[torch.rand(1, 1, 768)]
).images[0]

image.save('result_single.jpg')
```
 
</details>

<details>
<summary>Dual IP-Adapter</summary>

 ```python
import torch
from diffusers import FluxPipeline
from diffusers.utils import load_image

pipe: FluxPipeline = FluxPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-dev",
    torch_dtype=torch.bfloat16,
)

pipe.load_ip_adapter(
    ["XLabs-AI/flux-ip-adapter", "XLabs-AI/flux-ip-adapter"],
    weight_name=["ip_adapter.safetensors", "ip_adapter.safetensors"],
    image_encoder_pretrained_model_name_or_path="openai/clip-vit-large-patch14"
)

pipe.set_ip_adapter_scale([0.25, 0.25])
pipe.enable_sequential_cpu_offload()

image = pipe(
    width=1024,
    height=1024,
    prompt="A vintage picture of an astronaut in a starry sky",
    generator=torch.manual_seed(423),
    ip_adapter_image_embeds=[torch.rand(1,1,768), torch.rand(1,1,768)]
).images[0]

image.save('result_multi.jpg')
```

</details>


</details>


 <table>
        <tr>
            <th>No IP-Adapter</th>
            <th>Single IP-Adapter, 0.5 scale</th>
            <th>Dual IP-Adapter, 2x 0.25 scale </th>
        </tr>
        <tr>
            <td></td>
            <td></td>
            <td></td>
        </tr>
    </table>

**Review by hlky — APPROVED** [review]
Thanks @guiyrt!
